src/lucky/mime_type.cr

```diff
@@ -0,0 +1,94 @@
+# :nodoc:
+class Lucky::MimeType
+  alias Format = Symbol
+  alias AcceptHeaderSubstring = String
+  class_getter accept_header_formats = {} of AcceptHeaderSubstring => Format
+
+  register "application/json", :json
+  register "text/json", :json
+  register "application/jsonrequest", :json
+  register "application/xml", :xml
+  register "application/rss+xml", :rss
+  register "application/atom+xml", :atom
+  register "application/x-yaml", :yaml
+  register "text/yaml", :yaml
+  register "text/csv", :csv
+  register "text/css", :css
+  register "text/calendar", :ics
+  register "text/plain", :plain_text
+  register "multipart/form-data", :multipart_form
+  register "application/x-www-form-urlencoded", :url_encoded_form
+
+  def self.known_accept_headers : Array(String)
+    accept_header_formats.keys
+  end
+
+  def self.known_formats : Array(Symbol)
+    (accept_header_formats.values + [:html]).uniq
+  end
+
+  def self.registered?(format : Symbol) : Bool
+    known_formats.includes?(format)
+  end
+
+  def self.register(accept_header_substring : AcceptHeaderSubstring, format : Format) : Nil
+    accept_header_formats[accept_header_substring] = format
+  end
+
+  # :nodoc:
+  def self.determine_clients_desired_format(request, default_format : Symbol)
+    DetermineClientsDesiredFormat.new(request, default_format: default_format).call
+  end
+
+  private class DetermineClientsDesiredFormat
+    private getter request, default_format
+
+    def initialize(@request : HTTP::Request, @default_format : Symbol)
+    end
+
+    def call : Symbol?
+      accept = accept_header
+
+      if usable_accept_header? && accept
+        from_accept_header(accept)
+      elsif ajax?
+        :ajax
+      else
+        default_format
+      end
+    end
+
+    private def from_accept_header(accept : String) : Symbol?
+      Lucky::MimeType.accept_header_formats.find do |accept_header_substring, format|
+        accept.includes?(accept_header_substring)
+      end.try(&.[1])
```

Comment:
wouldn't `.first?` be the same as `[0]?`? If the array has 2 elements, then it would be `.last?`. But also, if that `find` block returned `nil`, then calling it directly on it would fail. 

`.try(&.[1])` isn't the "prettiest", but I think it's correct here.

Reply:
Oh pfff. You're totally right. Should have woken up a bit before I reviewed this.